pkg/component/controller/cplb_reconciler.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2024 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+	"time"
+
+	kubeutil "github.com/k0sproject/k0s/pkg/kubernetes"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// CPLBReconciler monitors the endpoints of the "kubernetes" service in the
+// "default" namespace. It notifies changes though the updateCh channel provided
+// in the constructor.
+type CPLBReconciler struct {
+	log             *logrus.Entry
+	kubeconfigPath  string
+	clientset       kubernetes.Interface
+	addresses       []string
+	mu              sync.RWMutex
+	updateCh        chan<- bool
+	watchCancelFunc context.CancelFunc
+}
+
+func NewCPLBReconciler(kubeconfigPath string, updateCh chan<- bool) *CPLBReconciler {
+	return &CPLBReconciler{
+		log:            logrus.WithField("component", "cplb-reconciler"),
+		kubeconfigPath: kubeconfigPath,
+		updateCh:       updateCh,
+	}
+}
+func (r *CPLBReconciler) getClientSet() error {
+	clientset, err := kubeutil.NewClientFromFile(r.kubeconfigPath)
+	if err != nil {
+		return err
+	}
+
+	r.clientset = clientset
+	return nil
+}
+
+func (r *CPLBReconciler) Start() error {
+	if err := r.getClientSet(); err != nil {
+		return fmt.Errorf("Failed to get clientset. Unable to : %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	r.watchCancelFunc = cancel
+	go r.watchAPIServers(ctx)
+
+	return nil
+}
+
+func (r *CPLBReconciler) Stop() {
+	r.log.Info("Stopping CPLB reconciler")
+	r.watchCancelFunc()
+}
+
+func (r *CPLBReconciler) watchAPIServers(ctx context.Context) {
+	// Before starting check that the API is actually responding
+	for {
```

Comment:
Maybe we could utilize our internal `watch` helper here? IMO would be much simpler and that already handles retries etc.

Reply:
Changed this as a separate commit. Can be squashed.